### PR TITLE
docs: v12 migration guide [LIBS-689]

### PR DESCRIPTION
**Pull request by KaiVandivier**
[LIBS-689](https://dhis2.atlassian.net/browse/LIBS-689)

- [X] Needs some React 18 info
- [x] Add last bit of React 18 info after final tests
- [x] Some information about env vars is pending some discussion

[LIBS-689]: https://dhis2.atlassian.net/browse/LIBS-689?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/migration/v12.md
Comment (kabaros) [inline docs/migration/v12.md:1]: it's a great guide @KaiVandivier - two comments:

- Shall we have a mention for the changes to better supprot TypeScript?
- Can we already have this merged with a disclaimer that it is still in alpha, so that people can give it a go?

File: docs/migration/v12.md
Comment (kabaros) [inline docs/migration/v12.md:276]: wil this be removed in the underlying Vite CLI?

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:1]: 1. Definitely! Are these the things you're thinking of?
    1. Bootstrap a typescript app with `yarn d2 app scripts init --typescript my-app-name`
    2. TypeScript is supported out of the box with Vite [Is this new, compared to CRA?]
    3. [Upcoming] Use `@dhis2/cli-style` runs type checking
4. Sure, we could merge this to master, but I'm kinda hoping the rest of alpha will be in master soon too haha

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:276]: Not sure which bullet point you mean haha, but here's an answer for each:
1. Env vars are under our control, but I'm actually rethinking this point, i.e. maybe we should keep variables on `process.env` -- Earlier, I thought `import.meta.env` was a pretty standard thing with ES6 and that we should migrate to that for env vars, but I think that's actually just a Vite pattern, so it might be "more standard" to keep env vars on `process.env`. If we do that, we can at least deprecate the `REACT_APP_` prefixes for the variable keys
2. (Probably obvious, but direction is also under our control)
3. For `--allowJsxInJs`, that's also under our control -- it's a flag that our own CLI reads, and adds a bunch of kinda complex Vite config in `makeViteConfig.mjs`. That additional config may or may not be supported in future versions of Vite, but I think we should remove _our_ `--allowJsxInJs` flag and config in the next `cli-app-scripts` version to make things simpler, standardized, and more performant

File: docs/scripts/migrate/js-to-jsx.md
Comment (kabaros) [inline docs/scripts/migrate/js-to-jsx.md:52]: maybe we can also mention that there might be small changes to the snapshots themselves as a result of the React 18 upgrade (spaces, and css modules names as far as I can tell) that needs to be fixed manually (i.e. update the snapshots and verify the changes are minimal and make sense)

File: docs/migration/v12.md
Comment (kabaros) [reply docs/migration/v12.md:1]: TS was also supported in CRA, but I think it's still worth it to mention it's supported in Vite .. we can also say that we had other improvements for types like adding types to UI library, and to-come-soon types for the API generated out of OpenAPI specs

File: docs/_sidebar.md
Comment (Topener) [inline docs/_sidebar.md:4]: This sidebar isn't used, you need to add it to the sidebar in the developer portal!

So after merging this PR you'd need to merge the developer portal PR as well to get it to show up.

Comment (KaiVandivier) [reply docs/_sidebar.md:4]: Ah okay, thanks for the reminder!

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:134]: Cool, added more to this section 👍

File: docs/scripts/migrate/js-to-jsx.md
Comment (KaiVandivier) [reply docs/scripts/migrate/js-to-jsx.md:52]: Added in both pages 🙂

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:1]: Added some text in the new & upcoming changes sections 👍

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:11]:
```suggestion
These are some things that you’ll see right away after upgrading:
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:13]:
```suggestion
-   It’s fast! Starting up an app is nearly instant, and building an app is about 3-4 times faster compared to Create React App (CRA)
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:19]:
```suggestion
    -   Hot Module Replacement (HMR) for code changes in the plugin will be as fast as in the app
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:21]: Is this command explained elsewhere? If so, can we link it? (or document it elsewhere if not)

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:26]: Can you link the vite docs about these tools?

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:32]: Just a note about this section, this will need adjusted once these features are either deployed or dismissed. I think it is fine to have it here but we must not forget. 

If there are already JIRA tickets/epics about these features, we should link it to this file as a todo.

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:78]: I feel wording could be changed slightly here for readability. Made a suggestion

```suggestion
With this update we also upgraded from React 16 to React 18. Upgrading to `@dhis2/cli-app-scripts` version 12 should generally a smooth process for the consuming apps. Most of the [changes](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) shouldn't affect most apps.
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:86]: Might be nice to use an admonition here.

```suggestion
:::info Note
Default params for functional components are computed on each render, so default values like arrays and objects as defaults may unintentionally retrigger `useEffect` and `useMemo` functions on each render. Instead, use a stable reference like so:
```

Should also add `:::` below the codeblock

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:94]:
```suggestion
:::

```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:117]: we'll need to do a little testing to see if this will render properly on the developer portal

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:159]:
```suggestion
`renderHook` is not a separate library anymore; it is part of the `@testing-library/react`, and it has [quite a different API](https://testing-library.com/docs/react-testing-library/api/#renderhook). 

:::warning Note
Some docs online still point to the old version, which can be confusing.
:::
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:281]:
```suggestion
:::tip
If it's not currently possible to migrate, there is an [unofficial React 18 adapter](https://github.com/cfaester/enzyme-adapter-react-18) that can help in the mean time. However, it is highly recommended this is a temporary solution and you should still schedule migrating away from Enzyme.
:::
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:287]: As users are in the process of migrating, this wording seems better.
```suggestion
With the migration away from Create React App, the formatting for environment variable names has gotten simpler.
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:291]:
```suggestion
The `REACT_APP_`-prefixed variables are now deprecated and will be removed in the next major version. However they will be kept around alongside the shorter `DHIS2_` variables in this version for backwards compatibility. 

For example, both `process.env.REACT_APP_DHIS2_MY_VAR` and `process.env.DHIS2_MY_VAR` will be available. 

:::tip
The `REACT_APP_` prefixed variables will be removed in the next version, so make sure to migrate to the shorter `DHIS2_` variable names.
:::
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:301]: Can the Maintenance app example be linked/explained?

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:303]: Does the mention of the Analytics app matter? It might confuse more than solve as the solution is a simple upgrade of a package. Maybe add the command to handle this upgrade instead?

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:307]:
```suggestion
The most likely and impactful breaking changes have been described in detail above. However this doesn't cover all the breaking changes and deprecations introduced in this upgrade. Below is the exhausted list. 
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:311]:
```suggestion
1. Node 18 or Node 20+ is required. Node 19 is not supported.
```

File: docs/migration/v12.md
Comment (Topener) [inline docs/migration/v12.md:322]: Can Login App example be linked/explained?

**Review by Topener — CHANGES_REQUESTED** [review]
First of all, amazing guide. It looks very detailed and exhaustive. 

I went through the guide with a **very nit-picky and oblivious point-of-view**, so I added a grand total of 21 comments and suggestions to make it easier to read and contextually make sense.

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md]: Added in [intro revisions](https://github.com/dhis2/app-platform/pull/884/commits/071f5dc8e22aafd60318ba69ac353da608326cd2), but added v12 since it's helpful to see the relevant package version

**Review by kabaros — APPROVED** [review]

File: docs/_sidebar.md
Comment (KaiVandivier) [reply docs/_sidebar.md:4]: Added to the side bar in this PR: https://github.com/dhis2/developer-portal/pull/167 (currently commented; I guess it will wait on this one to be merged)

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:117]: Ah, it doesn't seem to be working :( I copied this into the blog post PR to test:

Screenshot 2024-11-04 at 14 50 09

I'll look into docusaurus support for something similar

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md]: Moved this kind of language into the dev portal blog post: https://github.com/dhis2/developer-portal/pull/167

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md]: I think it's probably best just to remove the mention, which I've done -- I think the rest of the comment is enough info. I've also removed similar "noticed in ___" mentions

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md]: Removed this mention, as described in another comment

File: docs/migration/v12.md
Comment (KaiVandivier) [reply docs/migration/v12.md:117]: Ah, just needed to do this: https://github.com/dhis2/developer-portal/pull/167/commits/45c55d803b610be22fe30768662ad5d53b5cbcca

Screenshot 2024-11-04 at 16 19 48

File: docs/migration/v12.md
Comment (Topener) [reply docs/migration/v12.md:23]: Oh wow really? That's why I couldn't find it I guess

**Review by Topener — APPROVED** [review]
Awesome! Great work!

**Pull request merged**
